File: examples/search_engine_serper.py
#!/usr/bin/env python3
"""Google search via Serper API (engine=google_serper).

Requires:
  export SERPER_API_KEY=your-key
"""

import tarzi


def main():
    config = tarzi.Config.from_str(
        """
[search]
engine = "google_serper"
limit = 5
"""
    )
    # Prefer SERPER_API_KEY env over embedding keys in config.

    search_engine = tarzi.SearchEngine.from_config(config)
    query = "agentic AI frameworks"
    try:
        results = search_engine.search(query, 5)
        print(f"Found {len(results)} Serper results:")
        for i, result in enumerate(results):
            print(f"{i + 1}. {result.title}")
            print(f"   URL: {result.url}")
            print(f"   Snippet: {result.snippet}")
            print()
    except Exception as e:
        print(f"Serper search failed: {e}")
        print("Set SERPER_API_KEY (or search.api_key) and retry.")
    finally:
        search_engine.shutdown()


if __name__ == "__main__":
    main()

File: examples/search_engine_serper.rs
//! Google search via Serper API (`google_serper` / `serper`).
//!
//! Requires:
//!   export SERPER_API_KEY=your-key
//!
//! Run:
//!   cargo run --example search_engine_serper

use tarzi::{config::Config, search::SearchEngine};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt::init();

    let mut config = Config::load().unwrap_or_default();
    config.search.engine = "google_serper".to_string();
    config.search.limit = 5;
    // Prefer SERPER_API_KEY env; optional fallback:
    // config.search.api_key = Some("your-serper-api-key".to_string());

    let mut search_engine = SearchEngine::from_config(&config);
    let query = "agentic AI frameworks";

    match search_engine.search(query, config.search.limit).await {
        Ok(results) => {
            println!("Found {} Serper results:", results.len());
            for (i, result) in results.iter().enumerate() {
                println!("{}. {}", i + 1, result.title);
                println!("   URL: {}", result.url);
                println!("   Snippet: {}", result.snippet);
                println!();
            }
        }
        Err(e) => {
            eprintln!("Serper search failed: {e}");
            eprintln!("Set SERPER_API_KEY (or search.api_key) and retry.");
        }
    }

    search_engine.shutdown().await;
    Ok(())
}
